File: lib/src/main/java/com/jimulabs/mirrorsandbox/mockdata/MockData.java
package com.jimulabs.mirrorsandbox.mockdata;

import com.android.internal.util.Predicate;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Created by lintonye on 15-03-04.
 */
public class MockData {

    public String imageUrl(final ImageSize maxSize) {
        ImageGen.MockImage img = new ImageGen().tossOne(new Predicate<ImageGen.MockImage>() {
            @Override
            public boolean apply(ImageGen.MockImage mockImage) {
                return mockImage.width <= maxSize.width;
            }
        });
        return img.url;
    }

    public String personName() {
        return new PersonNameGen().tossOne();
    }

    public String phrase() {
        //TODO make proper phrase
        return new PersonNameGen().tossOne();
    }

    public String paragraph() {
        //TODO make proper paragraph
        return new PersonNameGen().tossOne();
    }

    public static abstract class MockDataGenerator<T> {
        private Random mRandom = new Random();

        T tossOne() {
            return tossOne(null);
        }

        T tossOne(Predicate<T> predicate) {
            List<T> filtered;
            if (predicate == null) {
                filtered = getSamples();
            } else {
                filtered = new ArrayList<>();
                for (T i : getSamples()) {
                    if (predicate.apply(i)) {
                        filtered.add(i);
                    }
                }
            }
            if (filtered.size() <= 0) {
                throw new IllegalArgumentException("No samples found that match the predicate");
            }
            int idx = mRandom.nextInt(filtered.size());
            return filtered.get(idx);
        }

        protected abstract List<T> getSamples();
    }
}

File: lib/src/main/kotlin/com/jimulabs/mirrorsandbox/Mockdata.kt
package com.jimulabs.mirrorsandbox

import java.util.Random
import kotlin.properties.Delegates

public class MockData {
    fun imageUrl(size: ImageSize): String {
        var img = ImageGen().tossOne { it.height <= size.size && it.width <= size.size }
        return img.url
    }

    fun personName(): String {
        return PersonNameGen().tossOne()
    }

    fun phrase(): String {
        //TODO make proper phrase
        return PersonNameGen().tossOne()
    }

    fun paragraph(): String {
        //TODO make proper paragraph
        return PersonNameGen().tossOne()
    }
}

public enum class ImageSize(val size: Int) {
    small : ImageSize(360)
    large : ImageSize(1080)
    hd : ImageSize(1980)
}

trait MockDataGenerator<T> {
    val samples: Stream<T>
    fun tossOne(filter: (T) -> Boolean = { true }): T {
        var filtered = samples.filter { filter.invoke(it) }
        var random = Random()
        var idx = random.nextInt(filtered.count());
        return filtered.elementAt(idx)
    }
}

